src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -1570,45 +1565,42 @@ public override long GetChars(int ordinal, long dataOffset, char[]? buffer, int
         if (length == 0)
             return 0;
 
-        var (bytesRead, charsRead) = DecodeChars(decoder, buffer, bufferOffset, length, ColumnLen - PosInColumn);
+        var (bytesRead, charsRead) = DecodeChars(decoder, buffer.AsSpan(bufferOffset, length), ColumnLen - PosInColumn);
 
         PosInColumn += bytesRead;
         _charPos += charsRead;
         return charsRead;
     }
 
-    (int BytesRead, int CharsRead) DecodeChars(Decoder decoder, char[] output, int outputOffset, int charCount, int byteCount)
+    (int BytesRead, int CharsRead) DecodeChars(Decoder decoder, Span<char> output, int byteCount)
     {
         var (bytesRead, charsRead) = (0, 0);
+        var outputLength = output.Length;
 
         while (true)
         {
             Buffer.Ensure(1); // Make sure we have at least some data
-
             var maxBytes = Math.Min(byteCount - bytesRead, Buffer.ReadBytesLeft);
-            decoder.Convert(Buffer.Buffer, Buffer.ReadPosition, maxBytes, output, outputOffset, charCount - charsRead, false,
-                out var bytesUsed, out var charsUsed, out _);
+            var bytes = Buffer.Buffer.AsSpan(Buffer.ReadPosition, maxBytes);
+            decoder.Convert(bytes, output, false, out var bytesUsed, out var charsUsed, out _);
             Buffer.ReadPosition += bytesUsed;
             bytesRead += bytesUsed;
             charsRead += charsUsed;
-            if (charsRead == charCount || bytesRead == byteCount)
+            if (charsRead == outputLength || bytesRead == byteCount)
                 break;
-            outputOffset += charsUsed;
-            Buffer.Clear();
+            output = output.Slice(charsUsed);
         }
 
         return (bytesRead, charsRead);
     }
 
     internal (int BytesSkipped, int CharsSkipped) SkipChars(Decoder decoder, int charCount, int byteCount)
     {
-        // TODO: Allocate on the stack with Span
-        if (_tempCharBuf == null)
-            _tempCharBuf = new char[1024];
+        Span<char> tempCharBuf = stackalloc char[1024];
```

Comment:
Allocating 1kb on a stack to me seems a bit extreme but maybe it's fine?

Reply:
Yeah 1kb is, but we might want to reduce it to 512 chars in that case :) 

There are quite some examples of 512 chars being stackalloced in the bcl 
https://github.com/search?q=repo%3Adotnet%2Fruntime%20StackAllocThreshold&type=code

Reply:
Changed